pkg/tests/end_to_end_tests/trace_tree_test.go

```diff
@@ -0,0 +1,381 @@
+    span_tags,
+    status_code,
+    resource_tags,
+    resource_schema_url_id
+)
+SELECT
+    'cac4d4cf-5451-4c1c-b8b6-2e45abe5f87f',
+    19,
+    null,
+    -1,
+    now(),
+    now(),
+    '{}'::jsonb::tag_map,
+    'STATUS_CODE_OK',
+    '{}'::jsonb::tag_map,
+    -1
+;`, schema.Trace)
+	exec, err = db.Exec(ctx, insertDecoy)
+	require.NoError(t, err, "Failed to insert decoy span: %v", err)
+	require.Equal(t, int64(1), exec.RowsAffected(), "Expected to insert 1 spans. Got: %v", exec.RowsAffected())
+}
+
+func testTraceTree(t *testing.T, ctx context.Context, db *pgxpool.Pool) {
+	const qry = `
+	select span_id, coalesce(parent_span_id, -1), lvl, path::text
+	from trace_tree('3dadb2bf-0035-433e-b74b-9075cc9260e8') 
+	order by span_id`
+
+	type result struct {
+		SpanId       int64
+		ParentSpanId int64
+		Level        int64
+		Path         string
+	}
+
+	t.Run("trace_tree", func(t *testing.T) {
+		expectedResults := []result{
+			{1, -1, 1, "{1}"},
+			{2, 1, 2, "{1,2}"},
+			{3, 1, 2, "{1,3}"},
+			{4, 1, 2, "{1,4}"},
+			{5, 2, 3, "{1,2,5}"},
+			{6, 2, 3, "{1,2,6}"},
+			{7, 2, 3, "{1,2,7}"},
+			{8, 3, 3, "{1,3,8}"},
+			{9, 3, 3, "{1,3,9}"},
+			{10, 3, 3, "{1,3,10}"},
+			{11, 4, 3, "{1,4,11}"},
+			{12, 4, 3, "{1,4,12}"},
+			{13, 6, 4, "{1,2,6,13}"},
+			{14, 8, 4, "{1,3,8,14}"},
+			{15, 12, 4, "{1,4,12,15}"},
+			{16, 13, 5, "{1,2,6,13,16}"},
+			{17, 13, 5, "{1,2,6,13,17}"},
+			{18, 15, 5, "{1,4,12,15,18}"},
+		}
+		rows, err := db.Query(ctx, qry)
+		require.NoError(t, err, "Error running trace_tree: %v", err)
+		defer rows.Close()
+
+		for _, expected := range expectedResults {
+			require.True(t, rows.Next(), "Expected a row for span id %v but got none", expected.SpanId)
+
+			var actual = result{}
+			err = rows.Scan(&actual.SpanId, &actual.ParentSpanId, &actual.Level, &actual.Path)
+			require.NoError(t, err, "Failed to scan row. Expected row for span id %v: %v", expected.SpanId, err)
+
+			require.Equal(t, expected, actual)
+		}
+		require.False(t, rows.Next(), "Found more result rows than expected")
+	})
+}
+
+func testUpstreamSpans(t *testing.T, ctx context.Context, db *pgxpool.Pool) {
+	type result struct {
+		SpanId       int64
+		ParentSpanId int64
+		Dist         int64
+		Path         string
+	}
+
+	eval := func(t *testing.T, qry string, expectedResults []result) {
+		rows, err := db.Query(ctx, qry)
+		require.NoError(t, err, "Error running upstream_spans: %v", err)
+		defer rows.Close()
+
+		for _, expected := range expectedResults {
+			require.True(t, rows.Next(), "Expected a row for span id %v but got none", expected.SpanId)
+
+			var actual = result{}
+			err = rows.Scan(&actual.SpanId, &actual.ParentSpanId, &actual.Dist, &actual.Path)
+			require.NoError(t, err, "Failed to scan row. Expected row for span id %v: %v", expected.SpanId, err)
+
+			require.Equal(t, expected, actual)
+		}
+		require.False(t, rows.Next(), "Found more result rows than expected")
+	}
+
+	t.Run("upstream_spans", func(t *testing.T) {
```

Comment:
There is another function `span_tree` that returns both upstream and downstream paths from the perspective of a given span. In the results of this function, a negative dist is the way to distinguish between the upstream and downstream paths.

So, for consistency, `upstream_spans` also returns negative dists.

Reply:
Ok so now I understand that negatives are introduces because of the `span_tree` function. Do you have a specific use-case in mind where negatives are helpful? I imagine that if user gets span tree and loads it in memory it is fairly easy to figure out parents or children by traversing the tree.

Reply:
The negative is an easy and efficient way to determine direction.

Reply:
`dist < 0 as is_upstream`
vs
`path[0] = span_id as is_upstream`